unable to subscribe to MarkerArray under ROS 2 Galactic

I'm trying to use the Marker plugin with a Mapviz install under ROS 2 Galactic to display a MarkerArray, but am getting the following error:
```
[mapviz-1] [ERROR] [1648669993.058815599] [mapviz]: Unhandled RCLError in Qt event loop: could not create subscription: failed to create topic, at /tmp/binarydeb/ros-galactic-rmw-cyclonedds-cpp-0.22.4/src/rmw_node.cpp:2552, at /tmp/binarydeb/ros-galactic-rcl-3.1.2/src/rcl/subscription.c:105
```
I had an earlier installation of Mapviz under ROS 2 Eloquent that was able to subscribe to the same topic and messages I'm attempting to visualize under Galactic. I reran `colcon build` on the publisher package under Galactic and confirmed that the messages were publishing with the command `ros2 topic echo /poi_marker_array visualization_msgs/msg/MarkerArray`. Any thoughts on whether this is a problem with Mapviz or with my publisher, or additional troubleshooting tips?

--- a/mapviz_plugins/src/marker_plugin.cpp
+++ b/mapviz_plugins/src/marker_plugin.cpp
@@ -123,14 +123,42 @@ namespace mapviz_plugins
     marker_array_sub_.reset();
     if (!topic_.empty())
     {
-      marker_sub_ = node_->create_subscription<visualization_msgs::msg::Marker>(
-          topic_,
-          rclcpp::QoS(100),
-          std::bind(&MarkerPlugin::handleMarker, this, std::placeholders::_1));
-      marker_array_sub_ = node_->create_subscription<visualization_msgs::msg::MarkerArray>(
-          topic_,
-          rclcpp::QoS(100),
-          std::bind(&MarkerPlugin::handleMarkerArray, this, std::placeholders::_1));
+      // ROS 2 does not allow for a simple way to subscribe to a general type of message (i.e. Marker and MarkerArray)
+      // That would require a way to de-serialize the data for mapviz to consume (based on message type)
+      // The code below checks for the topic type and subscribes in the appropriate manner
+
+      auto known_topics = node_->get_topic_names_and_types();
+      if (known_topics.count(topic_) > 0)
+      {
+        std::string topic_type = known_topics[topic_][0];
+        if (topic_type == "visualization_msgs/msg/Marker")
+        {
+          marker_sub_ = node_->create_subscription<visualization_msgs::msg::Marker>(
+              topic_,
+              rclcpp::QoS(100),
+              std::bind(&MarkerPlugin::handleMarker, this, std::placeholders::_1));
+        }
+        else if (topic_type == "visualization_msgs/msg/MarkerArray")
+        {
+          marker_array_sub_ = node_->create_subscription<visualization_msgs::msg::MarkerArray>(
+              topic_,
+              rclcpp::QoS(100),
+              std::bind(&MarkerPlugin::handleMarkerArray, this, std::placeholders::_1));
+        }
+        else
+        {
+          RCLCPP_ERROR(node_->get_logger(),
+              "Unable to subscribe to topic %s (unsupported type %s).",
+              topic_.c_str(), topic_type.c_str());
+          return;
+        }
+      }
+      else
+      {
+        RCLCPP_ERROR(node_->get_logger(),
+            "Unable to subscribe to topic %s (does not exist).", topic_.c_str());
+        return;
+      }
 
       RCLCPP_INFO(node_->get_logger(), "Subscribing to %s", topic_.c_str());
     }

